static/js/src/advantage/react/components/Subscriptions/SubscriptionCancel/SubscriptionCancel.tsx

```diff
@@ -1,57 +1,129 @@
 import {
-  Button,
+  ActionButton,
   ButtonAppearance,
-  Col,
-  Input,
   Modal,
-  Row,
+  Notification,
+  Spinner,
 } from "@canonical/react-components";
 import type { ModalProps } from "@canonical/react-components";
-import React from "react";
+import React, { useRef, useState } from "react";
+import * as Yup from "yup";
+import { useCancelContract, useUserSubscriptions } from "advantage/react/hooks";
+import { selectSubscriptionById } from "advantage/react/hooks/useUserSubscriptions";
+import { SelectedId } from "../Content/types";
+import { Formik, FormikProps } from "formik";
+import SubscriptionCancelFields from "./SubscriptionCancelFields";
+import { SubscriptionCancelValues } from "./SubscriptionCancelFields/SubscriptionCancelFields";
 
 type Props = {
-  onClose: ModalProps["close"];
+  selectedId?: SelectedId;
+  onCancelSuccess: () => void;
+  onClose: NonNullable<ModalProps["close"]>;
 };
 
-const SubscriptionCancel = ({ onClose }: Props) => {
+enum CancelError {
+  SubscriptionMissing = "subscription-missing",
+  Failed = "failed",
+}
+
+const CancelSchema = Yup.object().shape({
+  cancel: Yup.string()
+    .matches(/^cancel$/, "You must enter the correct confirmation text")
+    .required("The confirmation text is required"),
+});
+
+const generateError = (error: CancelError | null) => {
+  return (
+    <>
+      {error === CancelError.SubscriptionMissing ? (
+        <>
+          It could be that you have a pending payment that is blocking this
+          action.{" "}
+        </>
+      ) : null}
+      Contact{" "}
+      <a className="p-notification__action" href="/contact-us">
+        Canonical sales
+      </a>{" "}
+      if the problem persists.
+    </>
+  );
+};
+
+const SubscriptionCancel = ({
+  selectedId,
+  onCancelSuccess,
+  onClose,
+}: Props) => {
+  const [isValid, setIsValid] = useState(false);
+  const [error, setError] = useState<CancelError | null>(null);
+  const formRef = useRef<FormikProps<SubscriptionCancelValues> | null>(null);
+  const { data: subscription, isLoading } = useUserSubscriptions({
+    select: selectSubscriptionById(selectedId),
+  });
+  const cancelContract = useCancelContract(subscription);
+
   return (
     <div className="p-subscriptions__sticky-footer-modal">
       <Modal
         buttonRow={
-          <Button
+          <ActionButton
             appearance={ButtonAppearance.NEGATIVE}
             className="u-no-margin--bottom"
+            disabled={!isValid}
+            loading={cancelContract.isLoading}
+            success={cancelContract.isSuccess}
+            onClick={() => {
+              if (formRef.current) {
+                // Because this button appears outside of the form in the DOM
+                // this uses a ref to call the submit handler.
+                formRef.current.handleSubmit();
+              }
```

Comment:
You could probably wrap the modal with Formik as well to prevent having to use hacks such as this.

Reply:
The modal would have to be wrapped in both `Formik` and `form` so that the submit worked, however this results in changing the HTML structure of the modal in a way that affects the flex layout of the modal so this seemed like a simpler solution than having to override the Vanilla css styles.

Reply:
You'd only need to wrap the modal with `<Formik />`, leaving the `form` as is - no HTML structure changes required.

Then you could use the render function to pass down the handleSubmit:
```jsx
<Formik<SubscriptionCancelValues>>
  {({ handleSubmit }) => (
    <div className="p-subscriptions__sticky-footer-modal">
      <Modal
        buttonRow={
          <ActionButton
            onClick={handleSubmit}
```

Reply:
Ah yes I see what you're suggesting, that should work nicely.